From d48972d845acdbc95091b62d30e0a373bfc8ca37 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Sat, 9 May 2020 17:23:59 +0800
Subject: [PATCH] qwaylandxdgshell: Support switching between fullscreen and
 maximized

1/ The weston expected a commit to apply new window states.
2/ Switching between fullscreen and maximized requests unsetting the old
state firstly.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 .../shellintegration/xdg-shell/qwaylandxdgshell.cpp   | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell.cpp b/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell.cpp
index 0358eac9..1347082e 100644
--- a/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell.cpp
+++ b/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell.cpp
@@ -105,6 +105,9 @@ void QWaylandXdgSurface::Toplevel::applyConfigure()
         m_xdgSurface->m_window->resizeFromApplyConfigure(m_pending.size);
     }
 
+    // Trigger a update to commit new state
+    m_xdgSurface->m_window->window()->requestUpdate();
+
     m_xdgSurface->setSizeHints();
 
     m_applied = m_pending;
@@ -170,6 +173,11 @@ void QWaylandXdgSurface::Toplevel::requestWindowStates(Qt::WindowStates states)
     // Re-send what's different from the applied state
     Qt::WindowStates changedStates = m_applied.states ^ states;
 
+    if (changedStates & Qt::WindowFullScreen) {
+        if (!(states & Qt::WindowFullScreen))
+            unset_fullscreen();
+    }
+
     if (changedStates & Qt::WindowMaximized) {
         if (states & Qt::WindowMaximized)
             set_maximized();
@@ -183,8 +191,7 @@ void QWaylandXdgSurface::Toplevel::requestWindowStates(Qt::WindowStates states)
             if (screen) {
                 set_fullscreen(screen->output());
             }
-        } else
-            unset_fullscreen();
+        }
     }
 
     // Minimized state is not reported by the protocol, so always send it
-- 
2.34.1

